program/src/processor.rs

```diff
@@ -147,20 +147,31 @@ fn calculate_rewards_per_token(rewards: u64, token_supply: u64) -> Result<u128,
 // * The current accumulated rewards per token is always greater than or equal
 //   to the last accumulated rewards per token.
 // * The marginal rate multiplied by the token account balance does not exceed
-//   `u128::MAX`.
+//   `u64::MAX * 1e9` (scaling factor).
 //
 // The first condition is guaranteed by the program logic.
-// The second condition is guaranteed under the following constraints:
-// * The current supply of SOL is 500,000,000 * 1e9.
-// * The mint supply is less than or equal to 1e9.
 //
-// In the most extreme case, where the current rate is the supply of SOL and
-// the last rate is 0:
+// The second condition requires careful consideration of token distribution as
+// well as total rewards accrued by the system.
 //
-// * With a fixed SOL supply of 500,000,000 * 1e9, the maximum supported mint
-//   supply is approximately 4.5 * 1e9.
-// * With a fixed mint supply of 1e9, the maximum supported SOL supply is
-//   approximately 1.87 * 1e13 * 1e9.
+// Consider the current (at the time of this writing) circulating SOL supply of
+// 466,000,000 * 1e9 and a total token supply of 1e18.
+//
+// Some general benchmarks:
+//
+// * Given a marginal rate that is equal to 10% of the circulating supply of SOL
+//   (0.1 * 466,000,000 * 1e9), the maximum supported token account balance is
+//   approximately 395 * 1e9, or 0.0000395% of the token supply.
+// * Given a token account balance that is 10% of the token supply (0.1 * 1e18),
+//   the maximum supported marginal rate is approximately 184 SOL (184 * 1e9).
+//
+// In the most extreme cases:
+//
+// * Given a marginal rate that is equal to 100% of the circulating supply of
+//   SOL (466,000,000 * 1e9), the maximum supported token account balance is
+//   approximately 39.5 * 1e9, or 0.00000395% of the token supply.
+// * Given a token account balance that is 100% of the token supply (1e18), the
+//   maximum supported marginal rate is approximately 18.4 SOL (18.4 * 1e9).
```

Comment:
The biggest token account that we can imagine is the vault with all PAL tokens in the staking system, which will start with ~80% of all tokens. While the account should be updated pretty aggressively, if the system lands a couple of big arbs and gets 50 SOL quickly, are we screwed? A maximum tolerance of 1000 SOL would probably make me feel safe. What do you think?

As an option, we can do the calculation the other way around for really huge amounts. Ie. if `token_account_balance` or `marginal_rate` are greater than 10,000 * 10^9, then divide by `REWARDS_PER_TOKEN_SCALING_FACTOR` before multiplying to avoid overflow.

Reply:
Yeah, this is kind of what I was thinking, too. I’d feel safer if we could support a larger range of values here, especially if we know the vault will start out with that many tokens.

Reply:
My concern with flipping the order of these operations based on the size of the marginal rate or token account balance is that we lose precision for larger values, which is arguably where precision matters most.

Do you think it makes sense to store unharvested rewards as a `u128` on the holder rewards accounts, and only allow harvests up to `u64::MAX`? Obviously, there should never be a harvest even close to that, but using `u128` gives us more room in the calculation without losing that precision.

Some back-of-napkin math shows that we'd have the following new constraints, since `(current - last) * balance` now must be less than or equal to `u128::MAX`:

- Max marginal rate at 100% of token supply (`1e18`): `34_000_000_000` SOL.
- Max token account balance at 100% of circulating supply: `7.39 * 1e22`.

And then, for the vault:

- Max marginal rate at 80% of token supply (`0.8 * 1e18`): `42_500_000_000` SOL.

Reply:
I guess this seems like overkill for this system, although it should work quite fine.

If we don't want to go nuts, then I'll do as you suggested and use your values as a threshold, and just clearly document the function as losing precision beyond those values.

Reply:
Sorry, I forgot to descale my values, so you're right 34 SOL for 100% and 41.5 SOL for 80%. What I was alluding to was the fact that u128::MAX isn't our lowest bound.

> First of all, the constraint for the calculation actually requires (current - last) * balance to be less than or equal to u64::MAX * 1e9, which is significantly smaller than u128::MAX. This is because of the conversion to u64 that happens at the end of the calculate_eligible_rewards function.

I believe the values in my function documentation are also too high. They are the _scaled_ values, meaning we're super low on SOL threshold.

Reply:
Sorry for the confusion @joncinque but I think we're both fixating on the wrong thing actually. Or maybe just I am.

The function being prop-tested is `calculate_eligible_rewards`, which is evaluated on current/last rewards _per token_, NOT true rewards. Since either value is first calculated across the mint's supply (per token), then the bounds here are not reward amounts, but rewards _per token_ amounts.

```
rewards_per_token = (reward * e9) / mint_supply

eligible_rewards = (marginal_rewards_per_token * balance) / e9
```

Consider a reward of 1000 SOL (`1e12` lamports) and a mint supply of `e18`. Running this through `calculate_rewards_per_token` yields a reward _per token_ of 1000 lamports.

```
(e12 * e9) / e18 = 1000
```

We use `u128` to store this per token value since, over time, the accumulated amount (current, last) can get quite large.

Working backwards, if the maximum eligible reward is `u64::MAX * e9`, which is `1.8 * e28`, then we can set up the arithmetic like so:

```
1.8 * e28 = (max_marginal_rewards_per_token * balance) / e9
1.8 * e37 = max_marginal_rewards_per_token * balance
max_marginal_rewards_per_token = (1.8 * 1e37) / balance
```

Plugging in 100% of mint supply as the token account balance, we get `(1.8 * 1e37) / 1e18` which is `1.8 * e19` _rewards per token_.

Now we can work backwards to see what the maximum reward amount would be from that value. Conceptually this would be total rewards applied to the system that some holder may not have claimed (unharvested).

```
rewards_per_token = (reward * e9) / mint_supply
rewards_per_token * mint_supply = reward * e9
reward = (rewards_per_token * mint_supply) / e9
```

Plugging in `1.8 * e19` rewards per token and `e18` mint supply, we get a maximum accumulated _true_ reward of `(1.8 * e19 * e18) / e9` = `1.8 * e28` lamports, or `1.8 * e19` SOL.

So the maximum rewards that the system could have received over time that could break `calculate_rewards_per_token` is `1_800_000_000_000_000_000` SOL, which is significantly larger than both `u64::MAX` and the circulating supply.

Does that make sense? I'll need to adjust all of the documentation and probably the prop test constraints, but I think we're beyond covered here.

Reply:
Ah yeah that's a great point! If all of the supply of SOL (500M SOL) went to all PAL (1B PAL), that would only give a marginal rewards per token of 0.5 SOL, or 0.5 lamports per PAL token  :sweat_smile:

Could you add a test for this limit to make it clear that we're ok? I think this captures it:

```
    #[test]
    fn maximum_eligible_rewards() {
        // 1 lamport per token (not actually possible, but shows that we're ok)
        let maximum_marginal_rewards_per_token = REWARDS_PER_TOKEN_SCALING_FACTOR;
        let maximum_token_balance = 1_000_000_000 * 1_000_000_000; // 1 billion with 9 decimals
        let _ = calculate_eligible_rewards(
            maximum_marginal_rewards_per_token,
            0,
            maximum_token_balance,
        ).unwrap();
    }
```

Reply:
I did some more math on this, and I actually found some more interesting bounds. Check out my second-latest commit.